scalopus_catapult/src/catapult_recorder.cpp

```diff
@@ -0,0 +1,158 @@
+/*
+  Copyright (c) 2018-2019, 2021, Ivor Wanders
+  All rights reserved.
+
+  Redistribution and use in source and binary forms, with or without
+  modification, are permitted provided that the following conditions are met:
+
+  * Redistributions of source code must retain the above copyright notice, this
+    list of conditions and the following disclaimer.
+
+  * Redistributions in binary form must reproduce the above copyright notice,
+    this list of conditions and the following disclaimer in the documentation
+    and/or other materials provided with the distribution.
+
+  * Neither the name of the author nor the names of contributors may be used to
+    endorse or promote products derived from this software without specific
+    prior written permission.
+
+  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+  ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+#include "scalopus_catapult/catapult_recorder.h"
+#include <fstream>
+#include <sstream>
+
+namespace scalopus
+{
+void CatapultRecorder::loop()
+{
+  while(running_)
+  {
+    {
+      std::lock_guard<std::mutex> lock{source_mutex_};
+      for (auto& source : sources_)
+      {
+        source->work();
+      }
+    }
+
+    // Finally, block a bit just to prevent spinning.
+    // @TODO Think of a way to allow it to block here until there are messages or work needs to be done.
```

Comment:
Yeah, I know, all throughout scalopus we have spinners that just have a delay though. I didn't want to use a condition when I wrote it because I didn't want to risk spinning if there's just a single tracepoint, or forcing a spin for every tracepoint that's added.. Should probably be something where we spin when there's a high water mark in the buffers, but then it's tricky to also ensure they're flushed after a certain duration. Keeping this as is for now, we can grep for this comment if we ever clean it up.